design/baremetal-operator/maintenance-mode.md

```diff
@@ -0,0 +1,87 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Maintenance Mode flag
+
+## Status
+
+implementable
+
+## Summary
+
+Currently there is no way to prevent management of BaremetalHost resources
+after provisioning is completed.  The underlying Ironic API contains a
+maintenance-mode flag which could be used to disable management of
+BMH resources.
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+This may also be useful for temporary maintenance operations
+where you may wish for power management operations and polling
+to be stopped for specific hosts (this is the intended use if the
+underlying [Ironic API](https://docs.openstack.org/api-ref/baremetal/?expanded=set-maintenance-flag-detail#set-maintenance-flag)
+
+### Goals
+
+Add a maintenance flag to the BMH API which uses the underlying Ironic
+interface to disable post-deploy management of a host.
+
+### Non-Goals
+
+TBC
+
+## Proposal
+
+Introduce a new BMH field `maintenance` in the spec of the
+BaremetalHost object. This will be a boolean flag, which
+when set to `true` will put the underlying Ironic host
+into maintenance mode.
```

Comment:
It's not clear to me what actions are possible on a BMH that has maintenance=true.  Can I still provision/deprovision?
Can you please describe (either in comment or in the proposal) what the expected flow would be in the multi-tier deployment case?  Specifically, when should the BMH on the parent cluster have maintenance set to true and when to false?

Reply:
> It's not clear to me what actions are possible on a BMH that has maintenance=true. Can I still provision/deprovision?

No, the Ironic API causes all periodic tasks for that host to stop, and it's not a candidate for provisioning/deprovisioning (but you can delete it, without deprovisioning)

> Can you please describe (either in comment or in the proposal) what the expected flow would be in the multi-tier deployment case? Specifically, when should the BMH on the parent cluster have maintenance set to true and when to false?

Sure, lets see if we can discuss and agree the flow here, then I'll sync the result back into the doc, I'm thinking:

- User creates BMH resource(s) in parent cluster
- Provisioning of BMH triggered, which results in a running child cluster
- Parent cluster BMH resources all marked as `maintenance:  true` (or unmanaged, or paused, depending on the outcome of the other discussion threads)
- Child cluster BMH resources created, with BMC credentials, but marked `externallyProvisioned: true`

At this point, the physical hosts are owned by the child cluster BMH, but the inventory still exists in the parent cluster.

In the event that all of the child cluster hosts require reprovisioning, it would be necessary to remove the maintenance/unmanaged flag, or delete/re-create the BMH resource so that the process starts again at the beginning.

Reply:
OK that makes sense.  Basically it's up to the orchestration on the Hub to ensure that the Hub BMH is not in maintenance only when no Spoke BMH exists for that host.

Reply:
> OK that makes sense. Basically it's up to the orchestration on the Hub to ensure that the Hub BMH is not in maintenance only when no Spoke BMH exists for that host.

Right, that may be worth calling out explicity, e.g a non-goal is any coordination between the tiers of BMH resources, that has to be handled externally

Reply:
+1